### PR TITLE
Add CI and fix lint errors

### PR DESCRIPTION
**Pull request by kbressem**
## Summary
Adds basic CI for testing against multiple python versions and resolves linting errors

## CI

Installs Python and basic dependencies, then runs 
- lint with flake8
- formatting with black
- static type checking with mypy

After that, it continues to install the library, then runs tests (commented out, as we have no tests yet). 

I specified defaults in the `setup.cfg`for flake8, mypy and isort. Let's discuss if they make sense for you: 

- flake8 line length: 127 characters (flake8 default is 79, but this might be too short)
- black line length: 120 (flake8 longer allows us to sometimes exceed the 120 characters in doc-strings)
- max complexity: 10 

I was thinking of replacing the `setup.cfg` with a `pyproject.toml`, as this would allow to also configure black. However, I have not worked with toml files, yet.

## Fix lint errors
I have reformatted the code using black and fixed the few remaining flake8 errors. 

## Fix static type annotation
Fixing the mypy errors is quite challenging and I needed to do some code modifications. 

Noteworthy changes, which change functionality are: 

```python
if hasarg(self.criterion, "target") or hasarg(self.criterion.forward, "target"):  # type: ignore
    loss = self.criterion(pred, target.to(self._cast_target_dtype_to))
```
This allows for a criterion that is not a nn.Module. IIRC, statements are not evaluated full, so it wont raise an error if criterion does not has the `forward` attribute. Still, a bit hacky. 
In the second line, it would be possible to pass `None` as `target`. Currently I only see a large if statement as solution for this...
Overall, I am not sure of this specific function `_calculate_loss` add to the functionality of lighter or obscures problems with the loss function. Ideally the user should know what to pass to the loss, so we might afford to remove it completely. I have raised an issue for this. 

`LigterSystem.setup` and `LighterSystem._init_placeholders_for_dataloader_and_step_methods` required my to use multi-line mypy type-ignores. Maybe there is a better solution to this?

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# This workflow will install Python dependencies, runs lint, static typechecking and tests
+# with multiple versions of Python. 
+
+name: Continuous Integration
+
+on: push
```
Comment (surajpaib) [inline .github/workflows/ci.yml:6]: Do we want to run this on every push?

I think we should limit it to pull requests for master. Depends on what the purpose of this workflow would be at the end. If the intention is to use this to check for consistency, formatting and passing tests, then PR to main/master would be sufficient to trigger this imo

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        pip install setuptools wheel flake8 black mypy parameterized
+    - name: Lint with flake8
+      run: |
+        # stop the build if there are Python syntax errors or undefined names
+        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
+        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
+        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+    - name: black
+      run: |
+        black -l 120 --check lighter projects tests
```
Comment (surajpaib) [inline .github/workflows/ci.yml:34]: @ibro45 We previously used `yapf` for formatting, we should decide if we want to move to `black` or stick to `yapf` and proceed accordingly

File: setup.cfg [hunk 3]
```diff
@@ -68,4 +68,54 @@ fail-under = 8
 
 [pylint.typecheck]
 generated-members=torch.*, cv2.*
-good-names=x, y, z, i
\ No newline at end of file
+good-names=x, y, z, i
+
+[flake8]
```
Comment (surajpaib) [inline setup.cfg:73]: So I looked it up and seems that `flake8` doesn't have `pyproject.toml` support. However, `pylint` offers full support for it. Any reason to prefer flake over pylint? I've personally only used pylint so I am not aware of flake functionality. From what I can see in the config, we should be able to move everything over to pylint

**Review by surajpaib — COMMENTED** [review]
@kbressem Added some comments on the CI pipeline. 

I'm looking at a few repo templates currently to simplify the process for us. 

For example, code auto-formatting should probably be out of the CI and as a pre-commit hook or an auto-commit hook since at the end of the day, we want the code reformatted rather than throwing an error for not being formatted.

File: setup.cfg (hunk 3 above)
Comment (kbressem) [reply setup.cfg:73]: I have just used `flake8` until now, but I `pylint` is fine with me. Support of `pyproject.toml` is a big plus.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (kbressem) [reply .github/workflows/ci.yml:6]: Actions on open source repos are free, so I would recommend running them on every push. So you can see early if the CI got messed up before merging and it only takes a few minutes. Or do I miss some fundamental concepts here.

**Comment by kbressem** [comment]
> For example, code auto-formatting should probably be out of the CI and as a pre-commit hook or an auto-commit hook since at the end of the day, we want the code reformatted rather than throwing an error for not being formatted.

That is a great idea. Do you have any experience on setting up pre-commit hooks?

I was also thinking about including (CodeCov)[https://about.codecov.io], so we know if our tests are sufficient.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (kbressem) [reply .github/workflows/ci.yml:34]: I like black 😄.`monai` uses black as well (same configuration), so it might be easier to contribute stuff from `lighter` to `monai` of we use the same formatting. 

But similar to `flake8` I do not have a strong preference for black.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (surajpaib) [reply .github/workflows/ci.yml:6]: The running on PR only will achieve the same purpose. Any branch that submits a PR to main/master will run the CI. 

We can also run it on push for any branch but the downside might be that there might be quite a few commits pushed to a branch before a PR is made ready (In personal experience at least) and then we don't really need to run the CI (once we have more comprehensive testing) on each such commit as the purpose is to only check if everything looks okay before merging.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (surajpaib) [reply .github/workflows/ci.yml:34]: I agree, lets move to `black` then @ibro45

File: setup.cfg (hunk 3 above)
Comment (surajpaib) [reply setup.cfg:73]: Yes, in that case lets move to `pylint`. The `pyproject.toml` support will unify everything

**Comment by surajpaib** [comment]
Yes! Codecov would be great. Unfortunately I haven't used used it either. But I'm happy to take a look and see how we can set it up.

For the pre-commit hooks, I can set them up, no problem.

**Comment by kbressem** [comment]
@surajpaib given the larger changes which we also discussed in slack. Does it really help to change this PR or would it be better I close this and you open a new on with the project template, pre-commit hooks etc. I am happy to also help out there with setup.

**Comment by kbressem** [comment]
closed in favor of #17
